Gordon360/Controllers/RecIM/ActivitiesController.cs

```diff
@@ -0,0 +1,121 @@
+﻿using Gordon360.Models.CCT;
+using Gordon360.Services.RecIM;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Data.SqlClient;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Gordon360.Controllers.RecIM
+{
+    [Route("api/recim/[controller]")]
+    public class ActivitiesController : GordonControllerBase
+    {
+        private readonly IActivityService _activityService;
+
+        public ActivitiesController(IActivityService activityService)
+        {
+            _activityService = activityService;
+        }
+
+        ///<summary>Gets a list of all Activities</summary>
+        /// <returns>
+        /// All Existing Activities 
+        /// </returns>
+        [HttpGet]
+        [Route("")]
+        public ActionResult<IEnumerable<Activity>> GetAllLeagues()
+        {
+            var result = _activityService.GetActivities();
+            return Ok(result);
+        }
+
+        ///<summary>Gets a list of all Activities after given time</summary>
+        /// <returns>
+        /// All Existing Activities 
+        /// </returns>
+        [HttpGet]
+        [Route("{time}")]
+        public ActionResult<IEnumerable<Activity>> GetAllLeagues(DateTime time)
+        {
+            var result = _activityService.GetActivitiesByTime(time);
+            return Ok();
+        }
+
+        ///<summary>Gets a Activity object by ID number</summary>
+        /// <param name="activityID">League ID Number</param>
+        /// <returns>
+        /// Activity object
+        /// </returns>
+        [HttpGet]
+        [Route("{activityID}")]
+        public ActionResult<Activity> GetLeagueByID(int activityID)
```

Comment:
Hmmmm, regarding the first section, I'm not completely sure. Though, GetAllLeagues() has worked but I have yet to test the overrides. I'll test them in the next PR and update accordingly (most likely combining the the routes). Though it appears that I forgot to change the name League -> Activity in the controller.

Reply:
> I wonder if we even need the DateTime parameter route... What if we just made a route `/api/recim/leagues/active` and we can also throw in the optional `DateTime?` parameter if its needed (not sure what use case though?). I think this might actually be better because with calling a GET on `/api/recim/leagues` with a DateTime param, its pretty unclear what that param is doing. In this case we are looking for active leagues (narrowed by param or not). Does that makes sense?

That makes great sense to me. Would it be better to add `active` as an optional query param on the `/api/recim/leagues` route though? E.g. `/api/recim/leagues?active[=true if desired, but technically redundant]`. That would be a more RESTful approach, technically.

Reply:
I like the option of checking if "active" but one of the use cases I was thinking of would be for instances where we might want to see upcoming activities. We can make activities for the future and participants might want to be able to see if they want to join a tournament or what not in the future and plan their teams accordingly before the registration date opens. In that case, the activity would not be active, but we still might want to have the ability to show it.

Reply:
I think this is moving in the right direction, but it sounded like you might not have players and admins as two disjoint groups (i.e. you could have an admin who is also a player), and presumably someone who is both would have a use case for either version of the routes you gave, assuming that Admin views will exist separate from player views.

So maybe an optional `?admin` param (which would only be valid if the current user is an admin, of course) would be cleaner? Then you can distinguish `/activities?active&admin` from `/activities?active`, for example?

Reply:
The logic has the potential to get messy in the Controller, but I think [a switch expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression) would handle the valid cases pretty cleanly:
```csharp
var activities = (active, future, past, admin) switch
{
    (true, ..) => activitiesService.GetActive(),
    (, true, , true) => activitiesService.GetFutureForAdmin(),
    (, true, ..) => activitiesService.GetFuture(),
    // handle other cases
    _ => null
};

if (activities is null) return BadRequest();

// Other controller logic

Reply:
> /leagues gets all leagues (probably won't be used)
> /leagues?active gets all activities where the current date falls between RegistrationStartDate and ActivityEndDate
> /leagues?future gets all activities with RegistrationStartDate in the future
> if the authenticated user is an admin, this is all future activities
> otherwise this only fetches activities a certain distance in the future (again, either timeframe like month or current semester)
> leagues?past gets all leagues with ActivityEndDate in past
> [admin] fetches all past activities
> [player] fetches only past activities personally involved with

is easy enough to handle, my most recent commit has the following:
(all routes are prefixed by /recim/)
**activities** (where we have an optional "active" parameter which is defaulted to false, note that this is not a route parameter. I don't know if it's better to have it as a route param or not)
- this will return all activities period
- true will return all active activities based on our "completed" column (this will include all future activities as well as ongoing ones)
**activities?time={time}**
- this returns all activities which are active during the time period
^ the main reason for this, is I don't really know a use case where you would just want to see "every" single past or future activity. Rather, I think we could lazy load by semester/season.

-

Reply:
> The logic has the potential to get messy in the Controller, but I think [a switch expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression) would handle the valid cases pretty cleanly:
> 
> ```cs
> var activities = (active, future, past, admin) switch
> {
>     (true, ..) => activitiesService.GetActive(),
>     (, true, , true) => activitiesService.GetFutureForAdmin(),
>     (, true, ..) => activitiesService.GetFuture(),
>     // handle other cases
>     _ => null
> };
> 
> if (activities is null) return BadRequest();
> 
> // Other controller logic
> ```

I think this would probably be a cleaner solution, especially given what my current logic is.

Reply:
Just a thought: if it's true that you don't need to get ALL activities, you could make it so that the default behavior of `/activities` is to get only the currently active ones. That seems like a sensible and sane default. Then you would only have to handle the `date` and maybe `admin` filters, which would make things simpler.